src/systems/gimbal_controller/GimbalControllerPlugin.cc

```diff
@@ -0,0 +1,484 @@
+  rollTopic = this->model.Name(_ecm) + "/gimbal_roll_status";
+  this->rollPub= this->node.Advertise<ignition::msgs::StringMsg>(rollTopic);
+
+  // publish yaw status via gz transport
+  yawTopic = this->model.Name(_ecm) + "/gimbal_yaw_status";
+  this->yawPub= this->node.Advertise<ignition::msgs::StringMsg>(yawTopic);
+
+  ignmsg<< "GimbalControllerPluginPrivate::Init" <<std::endl;
+}
+
+//////////////////////////////////////////////////
+void GimbalControllerPluginPrivate::OnPitchStringMsg
+(const msgs::StringMsg &_msg)
+{
+	this->pitchCommand = atof(_msg.data().c_str());
+}
+
+//////////////////////////////////////////////////
+void GimbalControllerPluginPrivate::OnYawStringMsg
+(const msgs::StringMsg &_msg)
+{
+	this->yawCommand = atof(_msg.data().c_str());
+}
+
+//////////////////////////////////////////////////
+void GimbalControllerPluginPrivate::OnRollStringMsg
+(const msgs::StringMsg &_msg)
+{
+	this->rollCommand = atof(_msg.data().c_str());
+}
+
+//////////////////////////////////////////////////
+ignition::math::Vector3d GimbalControllerPluginPrivate::ThreeAxisRot(
+  double r11, double r12, double r21, double r31, double r32)
+{
+  return ignition::math::Vector3d(
+    atan2( r31, r32),
+    asin ( r21 ),
+    atan2( r11, r12));
+}
+
+//////////////////////////////////////////////////
+ignition::math::Vector3d GimbalControllerPluginPrivate::QtoZXY
+(const ignition::math::Quaterniond &_q)
+{
+  ignition::math::Vector3d result = this->ThreeAxisRot(
+    -2*(_q.X()*_q.Y() - _q.W()*_q.Z()),
+    _q.W()*_q.W() - _q.X()*_q.X() + _q.Y()*_q.Y() - _q.Z()*_q.Z(),
+    2*(_q.Y()*_q.Z() + _q.W()*_q.X()),
+    -2*(_q.X()*_q.Z() - _q.W()*_q.Y()),
+    _q.W()*_q.W() - _q.X()*_q.X() - _q.Y()*_q.Y() + _q.Z()*_q.Z());
+  return result;
+}
+
+//////////////////////////////////////////////////
+void GimbalControllerPlugin::PreUpdate(const ignition::gazebo::UpdateInfo &_info,
+    ignition::gazebo::EntityComponentManager &_ecm)
+{  if(!this->dataPtr->imuInitialized)
+   {
+      // Set unconditionally because we're only going to try this once.
+      this->dataPtr->imuInitialized = true;
+
+      auto imuEntity = _ecm.EntityByComponents(
+          components::Name(this->dataPtr->imuName),
+          components::Imu(),
+          components::ParentEntity(this->dataPtr->modelLink));
+
+      // Imu sensor topic 
+      std::string topic = scopedName(imuEntity, _ecm) + "/imu";
+
+      if(topic.empty())
+      {
+        ignerr << "imu_sensor [" << this->dataPtr->imuName
+               << "] not found, abort ArduPilot plugin." << "\n";
+        return;
+      }
+
+      // Subscribe to IMU sensor topic
+      this->dataPtr->node.Subscribe(topic, 
+        &GimbalControllerPluginPrivate::imuCb);
+    }
+    
+    if(!this->dataPtr->pitchJoint || !this->dataPtr->rollJoint 
+        || !this->dataPtr->yawJoint)
+      return;
+
+    if(_info.simTime < this->dataPtr->lastControllerUpdateTime)
+    {    
+       ignerr << "time reset event\n";
```

Comment:
Is this an error? If it is, we should be more verbose and explain to the user what will break in this case.